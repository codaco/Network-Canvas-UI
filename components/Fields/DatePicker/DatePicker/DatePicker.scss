.date-picker {

  &__panels {
    display: flex;
    width: 100%;
  }

  &__panel {
    background: pink;
    display: flex;
    flex-direction: column;
    flex: 0 1 33%;
  }

  &__panel-content {
    display: flex;
    flex-direction: row;
    max-height: 250px;
    background: green;
  }

  &__years {
    background: yellow;
    flex: 0 1 100%;
    overflow-y: scroll;
  }

  &__months,
  &__days {
    background: orange;
    flex: 0 1 100%;
    display: flex;
    flex-wrap: wrap;
  }

  &__year,
  &__month,
  &__day {
    display: flex;
    justify-content: center;
    align-items: center;
    height: 30px;
    cursor: pointer;

    &--active {
      background: red;
      border-radius: 50%;
    }
  }

  &__year {
    flex: 0 1 100%;
  }

  &__month {
    flex: 0 1 33%;
  }

  &__day {
    flex: 0 1 14%;
  }
}
