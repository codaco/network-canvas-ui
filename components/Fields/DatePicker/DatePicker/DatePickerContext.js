import React from 'react';

const DatePickerContext = React.createContext({
  onChange: () => {},
  min: null,
  max: null,
  date: {},
  type: null,
});

export default DatePickerContext;
