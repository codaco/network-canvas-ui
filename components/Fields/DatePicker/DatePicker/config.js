/* eslint-disable import/prefer-default-export */

export const DEFAULT_MIN_DATE = '1970-01-01';

export const DEFAULT_TYPE = 'full';

export const DATE_FORMATS = {
  full: 'yyyy-MM-dd',
  month: 'yyyy-MM',
  year: 'yyyy',
};
