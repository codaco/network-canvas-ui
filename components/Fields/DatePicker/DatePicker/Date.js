import { useContext } from 'react';
import DatePickerContext from './DatePickerContext';

const Date = ({ children }) => {
  const { onChange, date, set } = useContext(DatePickerContext);
  const props = {
    onChange,
    date,
    set,
  };
  return children(props);
};

export default Date;
