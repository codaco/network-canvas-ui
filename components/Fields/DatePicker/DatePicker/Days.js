import { useContext } from 'react';
import { range } from 'lodash';
import { DateTime } from 'luxon';
import DatePickerContext from './DatePickerContext';

const Days = ({ children }) => {
  const { onChange, date } = useContext(DatePickerContext);
  const days = range(1, DateTime.fromObject(date).daysInMonth + 1);
  const props = {
    days,
    day: date.day,
    onChange,
    date,
  };
  return children(props);
};

export default Days;
