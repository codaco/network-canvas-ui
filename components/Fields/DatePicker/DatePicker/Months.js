import { useContext } from 'react';
import { range } from 'lodash';
import DatePickerContext from './DatePickerContext';

const Months = ({ children }) => {
  const { onChange, date } = useContext(DatePickerContext);
  const months = range(1, 13);
  const props = {
    months,
    month: date.month,
    onChange,
    date,
  };
  return children(props);
};

export default Months;
