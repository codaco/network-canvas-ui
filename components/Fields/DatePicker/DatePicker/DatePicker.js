import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { DateTime } from 'luxon';
import DatePickerContext from './DatePickerContext';
import { DATE_FORMATS, DEFAULT_TYPE, DEFAULT_MIN_DATE } from './config';
import { now, isComplete } from './helpers';

/**
 * Get date object from an ISO string
 */
const getDate = (dateString) => {
  const { year, month, day } = dateString ?
    DateTime.fromISO(dateString).toObject() :
    {
      month: null,
      day: null,
      year: null,
    };
  return { year, month, day };
};

const DatePicker = ({
  children,
  ...props
}) => {
  const [pickerState, setPickerState] = useState({
    date: getDate(props.date),
  });

  const format = DATE_FORMATS[props.type];

  const min = props.min ?
    DateTime.fromISO(props.min) :
    DateTime.fromISO(DEFAULT_MIN_DATE);

  const max = props.max ?
    DateTime.fromISO(props.max) :
    now();

  const onChange = (values) => {
    const newDate = Object.assign({}, pickerState.date, values);

    setPickerState(state => ({
      ...state,
      date: newDate,
    }));

    if (isComplete(props.type)(newDate)) {
      const dateString = DateTime.fromObject(newDate).toFormat(format);
      props.onChange(dateString);
    }
  };

  const context = {
    onChange,
    min,
    max,
    type: props.type,
    ...pickerState,
  };

  return (
    <DatePickerContext.Provider value={context}>
      {children}
    </DatePickerContext.Provider>
  );
};

DatePicker.defaultProps = {
  children: null,
  type: DEFAULT_TYPE,
  min: null,
  max: null,
  onChange: () => {},
};

DatePicker.propTypes = {
  children: PropTypes.node,
  date: PropTypes.string.isRequired,
  type: PropTypes.string,
  min: PropTypes.string,
  max: PropTypes.string,
  onChange: PropTypes.func,
};

export default DatePicker;
