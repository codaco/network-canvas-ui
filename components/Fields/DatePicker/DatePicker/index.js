import { DateTime } from 'luxon';
import { DATE_FORMAT } from './DatePicker';

const now = DateTime.local().toFormat(DATE_FORMAT);

export { default as Years } from './Years';
export { default as Months } from './Months';
export { default as Days } from './Days';
export { default as Date } from './Date';
export { default as DatePicker } from './DatePicker';
export { DATE_FORMAT, now };
