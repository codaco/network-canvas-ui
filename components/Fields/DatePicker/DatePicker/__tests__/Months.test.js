/* eslint-env jest */

import React from 'react';
import { mount } from 'enzyme';
import DatePicker from '../DatePicker';
import Months from '../Months';

const getSubject = (props) => {
  const mockFunctionalComponent = jest.fn(() => null);

  mount((
    <DatePicker {...props}>
      <Months>{mockFunctionalComponent}</Months>
    </DatePicker>
  ));

  return mockFunctionalComponent;
};

describe('<Months>', () => {
  it('provides months', () => {
    const subject = getSubject();

    expect(subject.mock.calls[0][0]).toMatchObject({
      months: [
        1, 2, 3, 4, 5, 6, 7,
        8, 9, 10, 11, 12,
      ],
    });
  });
});
