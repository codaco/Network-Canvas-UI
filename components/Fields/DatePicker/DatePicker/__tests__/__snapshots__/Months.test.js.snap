// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Months> marks months as out of range 1`] = `
Object {
  "months": Array [
    Object {
      "isOutOfRange": true,
      "label": "January",
      "value": 1,
    },
    Object {
      "isOutOfRange": true,
      "label": "February",
      "value": 2,
    },
    Object {
      "isOutOfRange": true,
      "label": "March",
      "value": 3,
    },
    Object {
      "isOutOfRange": true,
      "label": "April",
      "value": 4,
    },
    Object {
      "label": "May",
      "value": 5,
    },
    Object {
      "label": "June",
      "value": 6,
    },
    Object {
      "label": "July",
      "value": 7,
    },
    Object {
      "label": "August",
      "value": 8,
    },
    Object {
      "isOutOfRange": true,
      "label": "September",
      "value": 9,
    },
    Object {
      "isOutOfRange": true,
      "label": "October",
      "value": 10,
    },
    Object {
      "isOutOfRange": true,
      "label": "November",
      "value": 11,
    },
    Object {
      "isOutOfRange": true,
      "label": "December",
      "value": 12,
    },
  ],
}
`;

exports[`<Months> provides months 1`] = `
Object {
  "months": Array [
    Object {
      "label": "January",
      "value": 1,
    },
    Object {
      "label": "February",
      "value": 2,
    },
    Object {
      "label": "March",
      "value": 3,
    },
    Object {
      "label": "April",
      "value": 4,
    },
    Object {
      "label": "May",
      "value": 5,
    },
    Object {
      "label": "June",
      "value": 6,
    },
    Object {
      "label": "July",
      "value": 7,
    },
    Object {
      "label": "August",
      "value": 8,
    },
    Object {
      "label": "September",
      "value": 9,
    },
    Object {
      "label": "October",
      "value": 10,
    },
    Object {
      "label": "November",
      "value": 11,
    },
    Object {
      "label": "December",
      "value": 12,
    },
  ],
}
`;
